KAFKA-6060: Add workload generation capabilities to Trogdor

cmccabe opened the pull request:
Previously, Trogdor only handled "Faults."  Now, Trogdor can handle
"Tasks" which may be either faults, or workloads to execute in the
background.

The Agent and Coordinator have been refactored from a
mutexes-and-condition-variables paradigm into a message passing
paradigm.  No locks are necessary, because only one thread can access
the task state or worker state.  This makes them a lot easier to reason
about.

The MockTime class can now handle mocking deferred message passing
(adding a message to an ExecutorService with a delay).  I added a
MockTimeTest.

MiniTrogdorCluster now starts up Agent and Coordinator classes in
paralle in order to minimize junit test time.

RPC messages now inherit from a common Message.java class.  This class
handles implementing serialization, equals, hashCode, etc.

Remove FaultSet, since it is no longer necessary.

Previously, if CoordinatorClient or AgentClient hit a networking
problem, they would throw an exception.  They now retry several times
before giving up.  Additionally, the REST RPCs to the Coordinator and
Agent have been changed to be idempotent.  If a response is lost, and
the request is resent, no harm will be done.

asfgit commented:
FAILURE
 No test results found.
--none--

cmccabe commented:
retest this please

cmccabe commented:
Testing @cmccabe

cmccabe commented:
Test failure was in `org.apache.kafka.streams.integration.ResetIntegrationTest.testReprocessingFromScratchAfterResetWithoutIntermediateUserTopic`, which is not related.  Re-pushing...

cmccabe commented:
Test failure is unrelated:
```
    org.apache.kafka.clients.NetworkClientTest.testConnectionDelayDisconnected
```

cmccabe commented:
Retest this please

cmccabe commented:
retest this please

cmccabe commented:
Test failed was: `org.apache.kafka.common.errors.GroupAuthorizationException: Not authorized to access group: group`, in     `kafka.api.SaslScramSslEndToEndAuthorizationTest.testTwoConsumersWithDifferentSaslCredentials` which is unrelated to the patch.

cmccabe commented:
retest this please

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/utils/Time.java:
```diff
@@ -53,4 +57,24 @@
+     * @return              A future which will complete when the callable is finished.
+     */
+    public <T> Future<T> schedule(final ScheduledExecutorService executor,
+                                  final Callable<T> callable, long delayMs);
```
I think we should try to do something similar to `kafka.utils.MockScheduler` keeping time and scheduler separate rather than add scheduling to `Time`. If the same code can be reused in `core` as well, that would be even better.

cmccabe replied:
I think the difference here is that we are reusing a standard Java interface, ScheduledExecutorService, rather than creating our own class, like KafkaScheduler.  I think the standard Java interface is more powerful and more familiar to most programmers.  But because it's a bigger interface, it's pretty hard to create a MockScheduledExecutorService.

In general, I'm not sure how to modify this to separate Scheduler from Time.  It seems like it might get pretty messy... we'd have to add Scheduler objects to the constructors of things.  But the Scheduler objects would have to take ownership of a thread created inside the thing, not passed from outside.

rajinisivaram replied:
An alternative may be to extend `Time` with a new interface in `tools` which does scheduling as well? Its usage would be the same as the current code, but the interface name could indicate that it is used for scheduling as well. And it could reuse the `Time` part of the implementation from `SystemTime` and `MockTime`.

cmccabe replied:
I'm looking into creating a MockScheduledExecutorService.  If I can do it, it will be more generally useful than the other options.  I have it almost written

rajinisivaram commented on tests/kafkatest/services/trogdor/trogdor.py:
```diff
@@ -253,3 +273,62 @@ def agent_class_name(self):
+#        error = task_state.get("error")
+#        if error is None or error == "":
+#            return None
+#        return error
```
Are these going to be removed?

cmccabe replied:
Removed

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/agent/AgentRestResource.java:
```diff
@@ -47,20 +49,19 @@ public void setAgent(Agent myAgent) {
-    public AgentFaultsResponse getAgentFaults() throws Throwable {
-        return agent().faults();
+    @PUT
+    @Path("/worker/create")
```
Perhaps this should be a `POST` to `/workers`?

cmccabe replied:
Good point.  I changed `/agent/worker/create` and `/coordinator/task/create` to be POST operations.

rajinisivaram replied:
Not sure you need `/create`. You could`POST` to `/agent/workers`? See comment below for `stop`.

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/agent/AgentRestResource.java:
```diff
@@ -47,20 +49,19 @@ public void setAgent(Agent myAgent) {
-    public Empty putAgentFault(CreateAgentFaultRequest request) throws Throwable {
-        agent().createFault(request);
-        return Empty.INSTANCE;
+    @Path("/worker/stop")
```
Perhaps this could be `DELETE` of `/workers/<id>`?

cmccabe replied:
I don't think this is a really a DELETE because the worker is still there, just stopped.  It's possible to imagine doing more with it in the future.

rajinisivaram replied:
In which case, it could be `PUT` to `/agent/workers/<id>`?

cmccabe replied:
I agree that it would be a little more "RESTy" to send different message types to `/agent/workers/$id`.  However, it's awkward to implement.  Right now, the APIs are "strongly typed" in the sense that when you send a message to `/agent/worker/stop`, you know what the format of the message is.  To implement something more complex, where you could send multiple messages to that endpoint, would be more difficult to implement with jaxrs + jetty.  Given that this is a purely internal and testing API for now, maybe we can do that later, once we've implemented more user-visible features?

rajinisivaram replied:
Yes, that is fine.

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/agent/WorkerManager.java:
```diff
@@ -0,0 +1,567 @@
+        CANCELLING,
+        RUNNING,
+        STOPPING,
+        DONE;
```
semicolon not needed

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/agent/WorkerManager.java:
```diff
@@ -0,0 +1,567 @@
+        /**
+         * The worker status.
+         */
+        private final AtomicReference<String> status = new AtomicReference<>("");
```
Where is this updated? And is it an `AtomicReference` because it can be updated/accessed from different threads?

cmccabe replied:
This is something which the worker can update in a background thread, asynchronously.  Currently, no workers are updating it, but in the future, they will update it with a status (like X% complete, etc.)

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/agent/WorkerManager.java:
```diff
@@ -0,0 +1,567 @@
+    /**
+     * Handles a worker halting.  Processed by the state change thread.
+     */
+    class HandleWorkerHalting implements Callable<Void> {
```
`HaltWorker` similar to `CreateWorker`? Or add `Handler` suffix to all?

cmccabe replied:
Hmm.  I think it's accurate to call this a "handler" since it is not initiating a halt.  It is handling a worker that halted itself for whatever reason.  In contrast, StopWorker actually does stop the worker (it's the backend for `/agent/workers/stop`)

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/common/ThreadUtils.java:
```diff
@@ -0,0 +1,46 @@
+ * Utilities for working with threads.
+ */
+public class ThreadUtils {
+    public static ThreadFactory createThreadFactory(final String pattern,
```
May be worth documenting pattern since it expects at most one `%d`.

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java:
```diff
@@ -46,27 +50,32 @@ public void setCoordinator(Coordinator myCoordinator) {
-    public CoordinatorFaultsResponse getCoordinatorFaults() throws Throwable {
-        return coordinator().getFaults();
+    @PUT
+    @Path("/task/create")
```
Similar to worker, could this be `POST` to `/tasks`?

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java:
```diff
@@ -46,27 +50,32 @@ public void setCoordinator(Coordinator myCoordinator) {
-        coordinator().createFault(request);
-        return Empty.INSTANCE;
+    @Path("/task/stop")
+    public StopTaskResponse stopTask(StopTaskRequest request) throws Throwable {
```
Similar to worker, could this be `DELETE` of `/tasks/<id>`?

cmccabe replied:
Similar to worker, the task is still there, so I'm not sure.  I have to think about it.

rajinisivaram replied:
Could be `PUT`? See comment on worker.

cmccabe replied:
See comment above

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java:
```diff
@@ -15,254 +15,315 @@
+ *
+ * RunTask actually starts the task which was created earlier.  This could
+ * happen an arbitrary amount of time after task creation (it is based on the
+ * task spec).  RunTask must operaate only on PENDING tasks... if the task has been
```
typo `operaate` -> `operate`

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java:
```diff
@@ -15,254 +15,315 @@
+ * RunTask asks the TaskController for a list of all the names of nodes
+ * affected by this task.
+ * If this list contains nodes we don't know about, or zero nodes, we
+ * transition direcctly to DONE state with an appropriate error set.
```
typo `direcctly` -> directly

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java:
```diff
@@ -0,0 +1,522 @@
+    }
+
+    public void waitForShutdown() throws ExecutionException, InterruptedException {
+        while (!executor.awaitTermination(1, TimeUnit.DAYS)) { }
```
Not sure we need the while loop since it waits for a day at least.

cmccabe replied:
It's there because people may call `waitForShutdown` prior to calling `beginShutdown`

rajinisivaram replied:
Makes sense, a comment would be useful.

cmccabe replied:
Added comments for beginShutdown and waitForShutdown

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/rest/JsonRestServer.java:
```diff
@@ -104,7 +110,7 @@ public void start(Object... resources) {
         jettyServer.setHandler(statsHandler);
         /* Needed for graceful shutdown as per `setStopTimeout` documentation */
-        jettyServer.setStopTimeout(GRACEFUL_SHUTDOWN_TIMEOUT_MS);
+        jettyServer.setStopTimeout(100);
```
100ms? Perhaps it was useful to have the constant since it at least made the unit clear.

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/rest/JsonRestServer.java:
```diff
@@ -119,17 +125,31 @@ public int port() {
-            jettyServer.destroy();
+    public void waitForShutdown() throws InterruptedException {
+        while (!shutdownExecutor.isShutdown()) {
+            shutdownExecutor.awaitTermination(1, TimeUnit.DAYS);
```
Not sure we need the while loop here either, just awaiting for a day should be sufficient?

cmccabe replied:
Same as above, people may call waitForShutdown prior to calling beginShutdown

rajinisivaram replied:
Again, comment would be useful.

cmccabe replied:
Added comments for beginShutdown and waitForShutdown

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/rest/JsonRestServer.java:
```diff
@@ -197,6 +217,26 @@ public void stop() {
+                Thread.sleep(10);
+            } else if (tries > 0) {
+                Thread.sleep(2);
+            }
```
Could just set the sleepTime and call `Thread.sleep` once?

cmccabe replied:
Hmm-- I might be misunderstanding this comment.  Wouldn't that be more verbose than what is here?

It would have to be like

            if (tries > 0) {
              if (tries > 1) {
                  sleepTime = 10;
              } else {
                 sleepTime = 2;
              }
              Thread.sleep(sleepTime);
           }

rajinisivaram replied:
I was thinking:
```
if (tries > 0)
    Thread.sleep(tries > 1 ? 10 : 2);
```

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/rest/WorkerReceiving.java:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * When we're in the process of sending a TaskSpec to the Agent, the Worker is regarded
+ * as being in SendingState.
```
Comment doesn't match the class?

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/rest/WorkerStopping.java:
```diff
@@ -19,51 +19,51 @@
 
 /**
- * A request to the Trogdor agent to create a fault.
+ * The state for a task which is being stopped on the agent.
```
The comment sounds confusing since it is the worker stopping?

rajinisivaram reviewed (commented):
@cmccabe I did a quick review and it is looking good. It was a very good idea to avoid locking. I have left a few comments for now. I will take another look later today.

rajinisivaram commented on tests/kafkatest/services/trogdor/trogdor.py:
```diff
@@ -226,24 +237,33 @@ def _coordinator_get(self, path, message):
+        :param spec:        The task spec.
         """
-        Create a new fault.
+        self._coordinator_put("task/create", { "id": id, "spec": spec.message()})
```
Should change to `POST` to match Java code changes.

cmccabe replied:
Good catch

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java:
```diff
@@ -0,0 +1,522 @@
+            for (NodeManager nodeManager : nodeManagers.values()) {
+                nodeManager.waitForShutdown();
+            }
+            executor.shutdownNow();
```
Isn't this the executor that is running the `Shutdown` task? Use `executor.shutdown()`?

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/fault/NetworkPartitionFaultController.java:
```diff
@@ -17,41 +17,30 @@
-public class NoOpFault extends AbstractFault {
-    private static final Logger log = LoggerFactory.getLogger(NoOpFault.class);
+public class NetworkPartitionFaultController implements TaskController {
+    private static final Logger log = LoggerFactory.getLogger(NetworkPartitionFaultController.class);
```
`log` not used

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/task/TaskSpec.java:
```diff
@@ -0,0 +1,105 @@
+     * point rounding.  (Objects serialized as JSON canonically contain only floating point numbers,
+     * because JavaScript did not support integers.)
+     */
+    public final static long MAX_TASK_DURATION = 1000000000000000L;
```
`MAX_TASK_DURATION_MS`?

rajinisivaram reviewed (commented):
Added a few more minor comments. It may also be useful to include some tests in `AgentTest/CoordinatorTest` where various actions throw exceptions.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/MockTime.java:
```diff
@@ -40,46 +57,108 @@ public MockTime(long autoTickMs) {
 
     @Override
-    public long milliseconds() {
+    public synchronized  long milliseconds() {
```
Did we make this synchronized due to the scheduling functionality?

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

rajinisivaram commented:
@cmccabe Can you fix the checkstyle errors since the builds are failing?

cmccabe commented:
Should be fixed

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

rajinisivaram commented on clients/out/test/resources/log4j.properties:
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-log4j.rootLogger=OFF, stdout
+log4j.rootLogger=TRACE, stdout
```
Perhaps didn't intend to checkin these changes? Actually I dont know why this file is even in github. Looking at the history, it has got me as the author, but it is not from my PR. Perhaps @ijuma checked it in inadvertently?

ijuma replied:
That's odd. https://github.com/apache/kafka/pull/4172

rajinisivaram replied:
I have merged #4172, thanks @ijuma.
@cmccabe Can you remove this change to avoid conflict? Thanks.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/utils/Timer.java:
```diff
@@ -0,0 +1,83 @@
+ *
+ * Implementations of this class should be thread-safe.
+ */
+public interface Timer {
```
Dont think this is being used, delete?

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/utils/MockScheduler.java:
```diff
@@ -0,0 +1,123 @@
+                future.complete(timeMs);
+            }
+            waiters.remove(entry.getKey());
+            completed++;
```
Why is `completed` count being tracked? It doesn't seem to be used anywhere.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/utils/MockTime.java:
```diff
@@ -24,6 +25,15 @@
+    /**
+     * Listeners which are waiting for time changes.
+     */
+    private final CopyOnWriteArrayList<MockTimeListener> listeners = new CopyOnWriteArrayList<>();
```
Is there a reason for this to be a `CopyOnWriteArrayList` as opposed to say a `ConcurrentLinkedQueue`?

cmccabe replied:
I would expect that an arraylist would perform slightly better than a linked list / queue when there were multiple elements.  However, the difference is small, especially since we currently only have one element.

rajinisivaram commented on clients/src/test/resources/log4j.properties:
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-log4j.rootLogger=OFF, stdout
+log4j.rootLogger=TRACE, stdout
```
Perhaps didn't intend to check this in?

rajinisivaram commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/MockTime.java:
```diff
@@ -18,7 +18,10 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ScheduledExecutorService;
```
Revert changes to this file?

rajinisivaram commented on core/src/test/scala/unit/kafka/producer/AsyncProducerTest.scala:
```diff
@@ -18,7 +18,7 @@
 
 import java.util.Properties
-import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.{Callable, Future, LinkedBlockingQueue, ScheduledExecutorService}
```
Revert changes to this file?

rajinisivaram commented on tests/kafkatest/tests/tools/trogdor_test.py:
```diff
@@ -12,15 +12,16 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import time
```
Not used?

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/common/Platform.java:
```diff
@@ -47,6 +51,7 @@ public static Platform parse(String curNodeName, String path) throws Exception {
             String platformName = platformNode.textValue();
             return Utils.newParameterizedInstance(platformName,
                 String.class, curNodeName,
+                Time.class, SystemTime.SYSTEM,
```
I see a unit test failure because these parameters don't match the constructor of `BasicPlatform`.

rajinisivaram reviewed (commented):
@cmccabe Had quick look through the code and left a few comments. The builds are still failing, so it will be good to get that fixed first.

cmccabe committed:
KAFKA-6060: Add workload generation capabilities to Trogdor

Previously, Trogdor only handled "Faults."  Now, Trogdor can handle
"Tasks" which may be either faults, or workloads to execute in the
background.

The Agent and Coordinator have been refactored from a
mutexes-and-condition-variables paradigm into a message passing
paradigm.  No locks are necessary, because only one thread can access
the task state or worker state.  This makes them a lot easier to reason
about.

The MockTime class can now handle mocking deferred message passing
(adding a message to an ExecutorService with a delay).  I added a
MockTimeTest.

MiniTrogdorCluster now starts up Agent and Coordinator classes in
paralle in order to minimize junit test time.

RPC messages now inherit from a common Message.java class.  This class
handles implementing serialization, equals, hashCode, etc.

Remove FaultSet, since it is no longer necessary.

Previously, if CoordinatorClient or AgentClient hit a networking
problem, they would throw an exception.  They now retry several times
before giving up.  Additionally, the REST RPCs to the Coordinator and
Agent have been changed to be idempotent.  If a response is lost, and
the request is resent, no harm will be done.

Add Scheduler interface.

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

rajinisivaram reviewed (approved):
@cmccabe Thank you for the PR, LGTM. Will merge after builds pass.

cmccabe commented:
jdk9 error is github integration error from jenkins, not the patch:
```
org.kohsuke.github.HttpException: Server returned HTTP response code: 201, message: 'Created' for URL: https://api.github.com/repos/apache/kafka/statuses/21585d7b0832c15fdbc234bea55c2531ef271e9c
	at org.kohsuke.github.Requester.parse(Requester.java:633)
	at org.kohsuke.github.Requester.parse(Requester.java:594)
	at org.kohsuke.github.Requester._to(Requester.java:272)
	at org.kohsuke.github.Requester.to(Requester.java:234)
	at org.kohsuke.github.GHRepository.createCommitStatus(GHRepository.java:1071)
	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.createCommitStatus(GhprbSimpleStatus.java:272)
	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.onBuildStart(GhprbSimpleStatus.java:185)
	at org.jenkinsci.plugins.ghprb.GhprbBuilds.onStarted(GhprbBuilds.java:130)
	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onStarted(GhprbBuildListener.java:18)
	at hudson.model.listeners.RunListener.fireStarted(RunListener.java:240)
	at hudson.model.Run.execute(Run.java:1733)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:419)
Caused by: java.io.IOException: Failed to deserialize {"url":"https://api.github.com/repos/apache/kafka/statuses/21585d7b0832c15fdbc234bea55c2531ef271e9c","id":4298877779,"state":"pending","description":"Build started sha1 is merged.","target_url":"https://builds.apache.org/job/kafka-pr-jdk9-scala2.12/2409/","context":"JDK 9 and Scala 2.12","created_at":"2017-11-02T23:09:55Z","updated_at":"2017-11-02T23:09:55Z","creator":{"login":"asfgit","id":1341245,"avatar_url":"https://avatars1.githubusercontent.com/u/1341245?v=4","gravatar_id":"","url":"https://api.github.com/users/asfgit","html_url":"https://github.com/asfgit","followers_url":"https://api.github.com/users/asfgit/followers","following_url":"https://api.github.com/users/asfgit/following{/other_user}","gists_url":"https://api.github.com/users/asfgit/gists{/gist_id}","starred_url":"https://api.github.com/users/asfgit/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/asfgit/subscriptions","organizations_url":"https://api.github.com/users/asfgit/orgs","repos_url":"https://api.github.com/users/asfgit/repos","events_url":"https://api.github.com/users/asfgit/events{/privacy}","received_events_url":"https://api.github.com/users/asfgit/received_events","type":"User","site_admin":false}}
```

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--
